### PR TITLE
[amp-social-share] Add SMS configuration and fix for iOS

### PR DESCRIPTION
**Pull request by RyanCopley**
# Turns out, iOS doesn't handle SMS as per spec. 

http://stackoverflow.com/a/19126326

So it appears as if  has decided ampersands are the separator of the QSP section of the sms URL. This small change allows native built in social sharing to SMS in a cross platform manner. If you all can think of a better way than doing string.replace, let me know, but I didn't want to dig into the urlReplacements code for fear of breaking everything. :)

# How it should have been

All platforms should have understood the format:

```
sms://{number}?body={prefill}
```

# How it is

Android supports the correct notation,  wants

```
sms://{number}&body={prefill}
```

I tried adding a "noop" param (e.g. `sms://{number}?noop=true&body={prefill}`), but that caused issues on Android. yay

# What does the IETF say

https://tools.ietf.org/html/rfc5724

```

2.2.  Formal Definition

   The URI scheme's keywords specified in the following syntax
   description are case-insensitive.  The syntax of an "sms" URI is
   formally described as follows, where the URI base syntax is taken
   from [RFC3986]:

  sms-uri        = scheme ":" sms-hier-part [ "?" sms-fields ]
  scheme         = "sms"
  sms-hier-part  = sms-recipient *( "," sms-recipient )
  sms-recipient  = telephone-subscriber ; defined in RFC 3966
  sms-fields     = sms-field *( "&" sms-field )
  sms-field      = sms-field-name "=" escaped-value
  sms-field-name = "body" / sms-field-ext ; "body" MUST only appear once
  sms-field-ext  = 1*( unreserved )
  escaped-value  = *( unreserved / pct-encoded ) ; defined in RFC 3986
```

lol

# What happens if you don't have the patch on mobilesafari

image1

# Try it

```html
<amp-social-share type="sms">Click me on an Android or iPhone, i'll work!</amp-social-share>
```

### CODE REVIEW

**Comment by aghassemi** [comment]
@RyanCopley Thanks for the PR Ryan.

Could you please 
- [ ] add a sample to `examples/social-share.amp.html` 
- [ ] update `amp-social-share.md` file
- [ ] Add a test for the iOS specific replacement of `body` to `test/test-amp-social-share.js` (use `'opens mailto: window in _top on iOS Safari'` test as reference)

Thank you!

**Comment by RyanCopley** [comment]
@aghassemi Absolutely

**Comment by aghassemi** [comment]
@spacedino Abby, do we have a SMS icon that matches the existing default ones we use ( https://ampbyexample.com/components/amp-social-share/ ). We also need to pick a default background color for SMS box too.

**Comment by RyanCopley** [comment]
```

.....................................................................................Error compiling extensions/amp-sticky-ad/1.0/amp-sticky-ad.js
ENOENT: no such file or directory, open 'build/cc/extensions_amp-sticky-ad_1.0_amp-sticky-ad.js'
```

Did I break this? It seems unrelated

**Comment by aghassemi** [comment]
@RyanCopley  nah just a flake in the build ( #9325 ). Restarting the build usually fixes it.

File: extensions/amp-social-share/0.1/amp-social-share.js [hunk 1]
```diff
@@ -92,9 +92,14 @@ class AmpSocialShare extends AMP.BaseElement {
       // mailto:, whatsapp: protocols breaks when opened in _blank on iOS Safari
       const isMailTo = /^mailto:$/.test(parseUrl(href).protocol);
       const isWhatsApp = /^whatsapp:$/.test(parseUrl(href).protocol);
+      const isSMS = /^sms:$/.test(parseUrl(href).protocol);
```
Comment (jridgewell) [inline extensions/amp-social-share/0.1/amp-social-share.js]: Why aren't these just equality checks?

File: extensions/amp-social-share/0.1/amp-social-share.js [hunk 2]
```diff
@@ -92,9 +92,14 @@ class AmpSocialShare extends AMP.BaseElement {
...
       const isIosSafari = this.platform_.isIos() && this.platform_.isSafari();
-      this.target_ = (isIosSafari && (isMailTo || isWhatsApp))
+      this.target_ = (isIosSafari && (isMailTo || isWhatsApp || isSMS))
           ? '_top' : '_blank';
+      if (isSMS && isIosSafari) {
+        // http://stackoverflow.com/a/19126326
+        this.href_ = this.href_.replace('?body', '&body');
```
Comment (jridgewell) [inline extensions/amp-social-share/0.1/amp-social-share.js]: Looks like we can use` sms://${NUMBER}?&body=${BODY}` (worked on iOS 10 and whatever android phone I have)

Comment (RyanCopley) [inline extensions/amp-social-share/0.1/amp-social-share.js]: I'll give this a shot

File: extensions/amp-social-share/0.1/amp-social-share.js (hunk 1 above)
Comment (RyanCopley) [inline extensions/amp-social-share/0.1/amp-social-share.js]: What are you suggesting I change? I just followed the conventions that already existed. It's probably best as a variable as I had to use it twice.

File: extensions/amp-social-share/0.1/amp-social-share.js (hunk 2 above)
Comment (RyanCopley) [inline extensions/amp-social-share/0.1/amp-social-share.js]: So, yes, `?&` works, but should it still make this kind of transform if it is not necessary on Android at all? Would you rather me do `if (isSMS) { this.href_ = this.href_.replace('?body', '?&body'); }`

File: extensions/amp-social-share/0.1/amp-social-share.js (hunk 1 above)
Comment (jridgewell) [inline extensions/amp-social-share/0.1/amp-social-share.js]: Just do `isSms = protocol == 'sms:'` (nit, capitalization)

File: extensions/amp-social-share/0.1/amp-social-share.js (hunk 2 above)
Comment (jridgewell) [inline extensions/amp-social-share/0.1/amp-social-share.js]: Yah, let's keep both paths the same. Could just be `if (isSms) { href = href.replace('?', '?&') }

File: extensions/amp-social-share/0.1/amp-social-share.js (hunk 1 above)
Comment (RyanCopley) [inline extensions/amp-social-share/0.1/amp-social-share.js]: Okay, done

File: extensions/amp-social-share/0.1/amp-social-share.js (hunk 2 above)
Comment (RyanCopley) [inline extensions/amp-social-share/0.1/amp-social-share.js]: Okay, done

**Comment by RyanCopley** [comment]
Should / can I add myself to the owners.yaml file?

File: extensions/amp-social-share/0.1/amp-social-share.js [hunk 3]
```diff
@@ -90,15 +90,17 @@ class AmpSocialShare extends AMP.BaseElement {
     urlReplacements.expandAsync(hrefWithVars).then(href => {
       this.href_ = href;
       // mailto:, whatsapp: protocols breaks when opened in _blank on iOS Safari
-      const isMailTo = /^mailto:$/.test(parseUrl(href).protocol);
-      const isWhatsApp = /^whatsapp:$/.test(parseUrl(href).protocol);
-      const isSMS = /^sms:$/.test(parseUrl(href).protocol);
+      const isMailTo = parseUrl(href).protocol === "mailto:";
```
Comment (jridgewell) [inline extensions/amp-social-share/0.1/amp-social-share.js]: Let's cache the protocol.

Comment (RyanCopley) [inline extensions/amp-social-share/0.1/amp-social-share.js]: oh ha, good idea

Comment (RyanCopley) [inline extensions/amp-social-share/0.1/amp-social-share.js]: Done

**Comment by aghassemi** [comment]
@RyanCopley yes to adding yourself to OWNER

@spacedino is creating a default icon for SMS, we would need to add that to `/optimized-svg-icons/amp-social-share-svgs.css` and `amp-social-share.css` before merging this PR. Stay tuned.

**Comment by RyanCopley** [comment]
@aghassemi awesome! I've added myself. You all have been wonderful!

**Comment by spacedino** [comment]
@aghassemi i dont know if we need another issue to track this but here are some options for a generic SMS icon, Version A and Version B

image

cc @ericlindley-g @cramforce

**Comment by aghassemi** [comment]
I like version B icon with version A color!

**Comment by spacedino** [comment]
[sms_version_b.svg.zip](https://github.com/ampproject/amphtml/files/1012078/sms_version_b.svg.zip)
purple hex: #5F067A

[sms_version_a.svg.zip](https://github.com/ampproject/amphtml/files/1012079/sms_version_a.svg.zip)
magenta hex: #CA2B63

ok! im ok with that too. here are the svgs

**Comment by spacedino** [comment]
live file.
[sms_icon.zip](https://github.com/ampproject/amphtml/files/1012081/sms_icon.zip)

ill file another ticket to adjust the size of the icons appropriately with in the boxes, right now some are too big.

**Comment by RyanCopley** [comment]
@aghassemi @spacedino Do I need to do anything here?

**Comment by aghassemi** [comment]
@RyanCopley if you like, we need to optimize the given SVG (version b) and add it to `/optimized-svg-icons/amp-social-share-svgs.css` and  also add `#CA2B63` as background color in `amp-social-share.css` before merging this PR. Up to you if you like to do it, we can merge and I can take care of that part after. Either is fine with me.

**Comment by RyanCopley** [comment]
I'll look into it tonight

**Comment by RyanCopley** [comment]
@aghassemi I'm an awful liar, i've already made the change and pushed 👍

**Comment by aghassemi** [comment]
Merged. Thanks for contributing.

**Pull request merged**